data/virt_autotest/fetch_logs_from_guest.sh

```diff
@@ -0,0 +1,260 @@
+	   fi
+	fi
+        
+	virsh start ${guest_domain}
+	if [[ ${ret1} -eq 0 && ${ret2} -eq 0 ]];then
+	   return 0
+	else
+	   return 1
+	fi
+}
+
+#Compress logs folder on local host which contains all logs from host and guest
+function compress_virt_logs_folder() {
+	local logs_folder=$1
+	local logs_root=${logs_folder/\//}
+        logs_root=${logs_root/\/*/}	
+
+	echo -e "tar -czvf /${logs_root}/virt_logs_all.tar.gz ${logs_folder}"
+        tar -czvf /${logs_root}/virt_logs_all.tar.gz ${logs_folder}
+	if [[ $? -eq 0 ]];then
+           echo -e "Successfully compressed ${logs_folder} to /${logs_root}/virt_logs_all.tar.gz"
+	   return 0
+       else
+           echo -e "Failed to ${logs_folder} to /${logs_root}/virt_logs_all.tar.gz"
+	   return 1
+	fi
+}
+
+#Usage and help info for the script
+help_usage(){
+	echo "script usage: $(basename $0) [-f \"Logs folder which contains logs collected(Can be omitted/Default to /tmp/virt_logs_residence)\"] [-g \"guests to be involved, for example, \"guest1 guest2 guest3\" or all or none(Can be omitted/Default to all)\"] [-e \"Extra folders or files to be fetched from guest, for example, \"log_file1 log_file2 log_folder1\"(Can be omitted/Default to nothing)\"] [-h help]"
+}
+
+fetch_logs_from_guest_log="/var/log/fetch_logs_from_guest.log"
+virt_guests_wanted=""
+virt_logs_folder=""
+virt_extra_logs_guest=""
+fetch_logs_from_guest_log_result=0
+rm -f -r ${fetch_logs_from_guest_log}
+
+#Parse input arguments
+while getopts 'l:g:e:h' OPTION; do
+  case "$OPTION" in
+    f)
+      virt_logs_folder="$OPTARG"
+      echo "Logs folder is ${virt_logs_folder}" | tee -a ${fetch_logs_from_guest_log}
+      ;;
+    g)
+      virt_guests_wanted="$OPTARG"
+      echo "The guests involved are ${virt_guests_wanted}" | tee -a ${fetch_logs_from_guest_log}
+      ;;
+    e)
+      virt_extra_logs_guest="$OPTARG"
+      virt_extra_logs_guest=(${virt_extra_logs_guest})
+      echo "The extra guest logs to be fetched are ${virt_extra_logs_guest[@]}" | tee -a ${fetch_logs_from_guest_log}
+      ;;
+    h)
+      help_usage | tee -a ${fetch_logs_from_guest_log}
+      exit 1
+      ;;
+    *)
+      help_usage | tee -a ${fetch_logs_from_guest_log}
+      exit 1
+      ;;
+  esac
+done
+shift "$(($OPTIND -1))"
+if [[ ${virt_logs_folder} == "" ]];then
+   virt_logs_folder="/tmp/virt_logs_residence"
+fi
+if [[ ${virt_guests_wanted} == "" ]];then
+   virt_guests_wanted="all"
+fi
+
+unset guest_hash_ipaddr
+declare -a guest_hash_ipaddr=""
+guest_domain_types="sles"
+guests_inactive_array=`virsh list --inactive | grep -E "${guest_domain_types}" | awk '{print $2}'`
+guest_domains_array=`virsh list  --all | grep -E "${guest_domain_types}" | awk '{print $2}'`
+guest_current=""
+guest_macaddresses_array=""
+guest_ipaddress="";
+guest_hash_index=0
+dhcpd_lease_file="/var/lib/dhcp/db/dhcpd.leases"
+
+#Establish virtual machine domain name and ip address mapping
+for guest_current in ${guest_domains_array[@]};do
```

Comment:
This part is nearly the same with the other PR 10788, can it be reused by putting in a lib?

Reply:
I have no intent to do that. I do not want to add another layer of lib thing to complicate things. I prefer to have more standalone and separate scripts. And actually, this is more about environment or running state, do not look too much like lib things to me.

Reply:
Well, line 232~239 can be wrapped as a library function IMO. Yes, it does bring one more layer and will bring a trouble thing to download also the library file before executing the script(from osd data directory) on SUT . 

However, when we take the plan to put new script code in openqa data directory, we will, definitely, sooner or later , need this lib file thing.

Reply:
This is not about bother only shell or bash programming. What you mean is to wrap something to another script in opneQA data directory. So the another script what you are going to call in this script is somehow unknown unless together with openQA and being determined at run time. Next you will need plenty of options to switch on and switch off. This is not good design from POV. One single script performs on single functionality and no inter-dependencies. Let alone this specific case, I do not want this thing to go to lib anyway.

Reply:
I agree with most of your opinions. Yes, it has disadvantages. The main disadvantage that I see, is when we want to execute some script(in openqa data), if it uses lib file(we should only have 1 lib file( for simplicity) for all future potential shared functions used by scripts that needs to be downloaded from data directory to SUT)), we have to download one more library file. That's it, no more others. When we take the plan to putting scripts in data, this is the main sacrifice that we have to accept. 
Regarding which function to put into this library and how should it be designed and called, it has nothing to do with the above design idea(create one library file for sharing functions reused among data scripts) and has nothing to do with the necessity of the design . It is within the regular programming rules. Among the rules, at least 1 main rule is to not duplicate code lines too much. Whenever starting copy lines, it shows library needs. 
BTW I do not fully agree with your point that one single script performs on single functionality and no inter-dependencies. Actually agree with the first part, not the latter part. There is no any complex project that can ensure no inter-dependencies at all. Instead nearly all projects have library files for reuse purpose.

Reply:
Of course if you think for your this specific code block, if putting into library, will need plenty of options to switch on and switch off, and it only duplicates twice, we do not win much by putting in library, I can accept it just duplicating as it is now. 

But still I want to highlight it to you that, maybe in future code changes, you can try to give higher attention to avoid the code duplication from very beginning. It is a very good programming habbit.

Reply:
I will consider all these and use my judgement to determine whether it should be go into lib. Sorry for this specific case, I really do not want to do that. I am not feeling obligatory to do that. By the way, this current imagined lib usage is crap.

Reply:
Of course not obligatory. Feel free to raise good library usage plan for the situation.